File: chesslib/position_factory.cpp
#include "position_factory.h"

#include "require.h"

#include <algorithm>

namespace chesslib {

Position createPosition(
    std::vector<PieceOnBoard> pieces,
    player_t playerToMove,
    OptionalColumn enPassantColumn,
    CastleOptions whiteCastleOptions,
    CastleOptions blackCastleOptions,
    uint32_t halfmoveClock,
    uint32_t fullmoveNumber)
{
    auto blackKingIt = std::find_if(pieces.begin(), pieces.end(),
        [] (const auto& piece) { return piece.pieceType == King && piece.player == Black; });

    auto whiteKingIt = std::find_if(pieces.begin(), pieces.end(),
        [] (const auto& piece) { return piece.pieceType == King && piece.player == White; });

    REQUIRE(blackKingIt != pieces.end() && whiteKingIt != pieces.end());

    auto position = Position(
        blackKingIt->square,
        whiteKingIt->square,
        playerToMove,
        halfmoveClock,
        fullmoveNumber);

    for (const auto& piece : pieces) {
        if (piece.pieceType != King) {
            position.addPiece(piece);
        }
    }

    position.setEnPassantColumn(enPassantColumn);
    position.setCastleOptions(White, whiteCastleOptions);
    position.setCastleOptions(Black, blackCastleOptions);

    position.optimizeCastleOptions();

    REQUIRE(position.isValid());
    return position;
}

}

File: android/chesslibkt/src/main/java/ru/maksimbulva/chesslibkt/position/PositionFactory.kt
package ru.maksimbulva.chesslibkt.position

import ru.maksimbulva.chesslibkt.Piece
import ru.maksimbulva.chesslibkt.Player
import ru.maksimbulva.chesslibkt.board.PieceOnBoard

object PositionFactory {

    fun createPosition(
        pieces: Collection<PieceOnBoard>,
        playerToMove: Player,
        enPassantColumn: Int?,
        castleOptions: Map<Player, CastleOptions>,
        halfmoveClock: Int,
        fullmoveNumber: Int
    ): Position {
        val blackKing = pieces.find { it.player == Player.Black && it.pieceType == Piece.King }!!
        val whiteKing = pieces.find { it.player == Player.White && it.pieceType == Piece.King }!!

        return Position(
            blackKingSquare = blackKing.square,
            whiteKingSquare = whiteKing.square,
            playerToMove = playerToMove,
            enPassantColumn = enPassantColumn,
            castleOptions = castleOptions,
            halfmoveClock = halfmoveClock,
            fullmoveNumber = fullmoveNumber
        ).apply {
            pieces.asSequence().filter { it.pieceType != Piece.King } .forEach { addPiece(it) }
            optimizeCastleOptions()
            require(isValid())
        }
    }
}
